language-plutus-core/stdlib/Language/PlutusCore/StdLib/Type.hs

```diff
@@ -633,39 +633,96 @@ packPatternFunctorBodyN ann dataName argVars patBodyN = do
         $ TyVar ann rec
 
 -- | Construct a data type out of pieces.
-getTyFix :: FromDataPieces ann (Type TyName ann)
-getTyFix ann name argVars patBodyN = do
+getPackedType :: FromDataPieces ann (Type TyName ann)
+getPackedType ann dataName argVars patBodyN = do
     withSpine <- getWithSpine ann argVars
-    withSpine . TyIFix ann <$> packPatternFunctorBodyN ann name argVars patBodyN
+    withSpine . TyIFix ann <$> packPatternFunctorBodyN ann dataName argVars patBodyN
 
 -- | An auxiliary type for returning a polymorphic @wrap@. Haskell's support for impredicative
 -- polymorphism isn't good enough to do without this.
-newtype PolyWrap ann =
-  PolyWrap { unPolyWrap :: forall term . TermLike term TyName Name
-                        => [Type TyName ann] -> term ann -> term ann
-           }
+newtype PolyWrap ann = PolyWrap
+    (forall term. TermLike term TyName Name => [Type TyName ann] -> term ann -> term ann)
 
 -- | Make a generic @wrap@ that takes a spine of type arguments and the rest of a term, packs
 -- the spine using the CPS trick and passes the spine and the term to 'IWrap' along with a 1-ary
 -- pattern functor constructed from pieces of a data type passed as arguments to 'getWrap'.
-getWrap :: FromDataPieces ann (PolyWrap ann)
-getWrap ann name argVars patBody = do
-    pat1 <- packPatternFunctorBodyN ann name argVars patBody
+getPackedWrap :: FromDataPieces ann (PolyWrap ann)
+getPackedWrap ann dataName argVars patBodyN = do
+    pat1 <- packPatternFunctorBodyN ann dataName argVars patBodyN
     toSpine <- getToSpine ann
     let instVar v ty = TyDecl ann ty $ tyVarDeclKind v
     return $ PolyWrap $ \args ->
         let argVarsLen = length argVars
             argsLen = length args
             in if argVarsLen == argsLen
                 then iWrap ann pat1 . toSpine $ zipWith instVar argVars args
-                else throw . IndicesLengthsMismatchException argVarsLen argsLen $ void name
+                else throw . IndicesLengthsMismatchException argVarsLen argsLen $ void dataName
+
+-- Note that the pattern functor of the resulting type is always the same, it's the index
+-- that differs.
+-- | Construct a 'RecursiveType' by passing a 0-ary pattern functor to 'TyIFix' and 'IWrap'
+-- /as an index/.
+makeRecursiveType0
+    :: ann              -- ^ An annotation placed everywhere we do not have annotations.
+    -> TyName ann       -- ^ The name of the data type being defined.
+    -> Type TyName ann  -- ^ The body of the pattern functor.
+    -> Quote (RecursiveType ann)
+makeRecursiveType0 ann dataName patBody0 = do
+    rec <- freshTyName ann "rec"
+    f   <- freshTyName ann "f"
+    let indKind = KindArrow ann (Type ann) $ Type ann
```

Comment:
Not sure what "ind" indicates here.

Reply:
"Index". The `i` part of `ifix`. `indKind` is the kind of `ind`, which is later used in `TyIFix ann pat1 ind`. I think this naming is pretty descriptive.

Reply:
Let's have `indexKind`. Empirically: I couldn't figure out what it was, my brain kept suggesting "indirect kind"!

Reply:
I see that everywhere else I use `ifix pat arg`. Forgot my own naming conventions. I'll thus change `indKind` to `argKind` and `ind` to `arg`, but I think we need a separate refactoring that replaces `arg` by `ind` everywhere in the context of `ifix`.

I prefer to avoid clashes with at the very least `Prelude` and `Data.List` and `indexKind` entails `index`, which is a clash. In general, I think having `ind` introduced somewhere near `pat` throughout the codebase is pretty unambiguous.

Reply:
Okay, how about `ix`? That's a more commonly used abbreviation for `index`.

Reply:
> Okay, how about `ix`?

Okay.

Made it `arg` instead of `ind` as everywhere else in the codebase. I'd be fine we a separate refactoring turning it into `ix`.